src/libappimage/utils/IconHandle.cpp

```diff
@@ -0,0 +1,417 @@
+                    CAIRO_STATUS_INVALID_DASH,
+                    CAIRO_STATUS_INVALID_DSC_COMMENT,
+                    CAIRO_STATUS_INVALID_INDEX,
+                    CAIRO_STATUS_CLIP_NOT_REPRESENTABLE,
+                    CAIRO_STATUS_TEMP_FILE_ERROR,
+                    CAIRO_STATUS_INVALID_STRIDE,
+                    CAIRO_STATUS_FONT_TYPE_MISMATCH,
+                    CAIRO_STATUS_USER_FONT_IMMUTABLE,
+                    CAIRO_STATUS_USER_FONT_ERROR,
+                    CAIRO_STATUS_NEGATIVE_COUNT,
+                    CAIRO_STATUS_INVALID_CLUSTERS,
+                    CAIRO_STATUS_INVALID_SLANT,
+                    CAIRO_STATUS_INVALID_WEIGHT,
+                    CAIRO_STATUS_INVALID_SIZE,
+                    CAIRO_STATUS_USER_FONT_NOT_IMPLEMENTED,
+                    CAIRO_STATUS_DEVICE_TYPE_MISMATCH,
+                    CAIRO_STATUS_DEVICE_ERROR,
+                    CAIRO_STATUS_INVALID_MESH_CONSTRUCTION,
+                    CAIRO_STATUS_DEVICE_FINISHED,
+                    CAIRO_STATUS_JBIG2_GLOBAL_MISSING,
+                    CAIRO_STATUS_PNG_ERROR,
+                    CAIRO_STATUS_FREETYPE_ERROR,
+                    CAIRO_STATUS_WIN32_GDI_ERROR,
+                    CAIRO_STATUS_TAG_ERROR,
+
+                    CAIRO_STATUS_LAST_STATUS
+                } cairo_status_t;
+
+                typedef void* (* cairo_image_surface_create_t)(int format, int width, int height);
+
+                typedef void* (* cairo_create_t)(void* target);
+
+                typedef int (* cairo_write_func_t)(void* closure, const unsigned char* data,
+                                                   unsigned int length);
+
+                typedef int (* cairo_surface_write_to_png_stream_t)(void* surface, cairo_write_func_t write_func,
+                                                                    void* closure);
+
+                typedef void (* cairo_destroy_t)(void* cr);
+
+                typedef void (* cairo_surface_destroy_t)(void* surface);
+
+                typedef void (* cairo_scale_t)(void* cr, double sx, double sy);
+
+                typedef cairo_status_t (* cairo_read_func_t)(void* closure,
+                                                             unsigned char* data,
+                                                             unsigned int length);
+
+                typedef void* (* cairo_image_surface_create_from_png_stream_t)(cairo_read_func_t read_func,
+                                                                               void* closure);
+
+                typedef void* (* cairo_image_surface_create_from_png_t )(const char* filename);
+
+                typedef int (* cairo_surface_status_t)(void* surface);
+
+                typedef int (* cairo_image_surface_get_height_t)(void* surface);
+
+                typedef const char* (* cairo_status_to_string_t)(int status);
+
+
+            public:
+                CairoHandle() : DLHandle("libcairo.so.2") {
```

Comment:
They are not meant to be reusable, they are made to fulfill the libappimage image manipulation requirements

Reply:
Everything should be designed so it can be reused. You never know what future brings. IF we ever need image manipulation again, we have to do the exact same work twice, because refactoring this at a later stage is very likely to be more work than doing it right now.

Reply:
In such a case, we will use `Cairo` and `rsvg` properly instead of loading them on demand at runtime.

Design to be ready for change is good practice. But also it's a trapif not done with care, it's not possible predict all the changes that will be required in the future. You have to design for the changes that you know that will happen. Otherwise it will be a waste of time and effort.

If you know of a use case where this will be required please let me know and we do the refactorization in a separated issue.

Reply:
The functionality implemented privately (i.e., without any realistic chance to access from outside) is _way_ too large to be tested properly by unit tests. I insist on moving it into separate classes. We can do it at a later stage, though. It's also pretty low prio. We should open an issue for it.